app/scripts/migrations/120.3.ts

```diff
@@ -0,0 +1,110 @@
+import { RuntimeObject, hasProperty, isObject } from '@metamask/utils';
+import { cloneDeep } from 'lodash';
+import log from 'loglevel';
+
+type VersionedData = {
+  meta: { version: number };
+  data: Record<string, unknown>;
+};
+
+export const version = 120.3;
+
+/**
+ * This migration trims the size of any large transaction histories. This will
+ * result in some loss of information, but the impact is minor. The lost data
+ * is only used in the "Activity log" on the transaction details page.
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly
+ * what we persist to dist.
+ * @param originalVersionedData.meta - State metadata.
+ * @param originalVersionedData.meta.version - The current state version.
+ * @param originalVersionedData.data - The persisted MetaMask state, keyed by
+ * controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(
+  originalVersionedData: VersionedData,
+): Promise<VersionedData> {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  transformState(versionedData.data);
+  return versionedData;
+}
+
+function transformState(state: Record<string, unknown>) {
+  if (!hasProperty(state, 'TransactionController')) {
+    log.warn(`Migration ${version}: Missing TransactionController state`);
+    return state;
+  } else if (!isObject(state.TransactionController)) {
+    global.sentry?.captureException(
+      `Migration ${version}: Invalid TransactionController state of type '${typeof state.TransactionController}'`,
+    );
+    return state;
+  }
+
+  const transactionControllerState = state.TransactionController;
+
+  if (!hasProperty(transactionControllerState, 'transactions')) {
+    log.warn(
+      `Migration ${version}: Missing TransactionController transactions`,
+    );
+    return state;
+  } else if (!Array.isArray(transactionControllerState.transactions)) {
+    global.sentry?.captureException(
+      `Migration ${version}: Invalid TransactionController transactions state of type '${typeof transactionControllerState.transactions}'`,
+    );
+    return state;
+  }
+
+  const validTransactions =
+    transactionControllerState.transactions.filter(isObject);
+  if (
+    transactionControllerState.transactions.length !== validTransactions.length
+  ) {
+    const invalidTransaction = transactionControllerState.transactions.find(
+      (transaction) => !isObject(transaction),
+    );
+    global.sentry?.captureException(
+      `Migration ${version}: Invalid transaction of type '${typeof invalidTransaction}'`,
+    );
+    return state;
+  }
+
+  const validHistoryTransactions = validTransactions.filter(
+    hasValidTransactionHistory,
+  );
+  if (validHistoryTransactions.length !== validTransactions.length) {
+    const invalidTransaction = validTransactions.find(
```

Comment:
Rather than this repeated pattern of filtering and finding if the lengths don't match, would it be simpler to just iterate once and check each transaction for validity as we go?

We could potentially handle multiple invalid transactions easier also?

Reply:
My goal was to ensure that the migration was either completed in full, or made zero state mutations. The idea was that this would make any follow-ups simpler to think about. That's why I try to do all state validation before all state mutation. Validating as we go would leave us in a very confusing state in some cases.

In this specific case, I can't imagine validating and mutating iteratively would cause problems because the mutation we're making here is rather simple, no structural changes. But it's a habit I try to follow on all migrations because it makes a very big difference in some other cases.

As far as how the validation was done, that was mainly motivated by TypeScript inference. I tried to validate things one at a time, building up the type for the variable with each check.